common/xdc/history_event_uitl.go

```diff
@@ -0,0 +1,669 @@
+			historyEvent.ExternalWorkflowExecutionCancelRequestedEventAttributes = &shared.ExternalWorkflowExecutionCancelRequestedEventAttributes{
+				InitiatedEventId: common.Int64Ptr(requestExternalWorkflowCanceledEventID),
+				Domain:           common.StringPtr("domain"),
+				WorkflowExecution: &shared.WorkflowExecution{
+					WorkflowId: common.StringPtr("external_wid"),
+					RunId:      common.StringPtr("external_rid"),
+				},
+			}
+		}
+		eventID++
+		historyEvents = append(historyEvents, historyEvent)
+	}
+	return historyEvents
+}
+
+func getDefaultHistoryEvent(eventID int64) *shared.HistoryEvent {
+	return &shared.HistoryEvent{
+		EventId:   common.Int64Ptr(eventID),
+		Timestamp: common.Int64Ptr(time.Now().Unix()),
+		TaskId:    common.Int64Ptr(common.EmptyEventTaskID),
+		Version:   common.Int64Ptr(common.EmptyVersion),
+	}
+}
+
+func doBatchEvents(events []*shared.HistoryEvent) []*shared.History {
+	history := make([]*shared.History, 0)
+	buffer := make([]*shared.HistoryEvent, 0)
+	noPendingDecisionTask := true
+	for _, event := range events {
+		buffer = append(buffer, event)
+		switch *event.EventType {
+		case shared.EventTypeDecisionTaskStarted,
+			shared.EventTypeDecisionTaskScheduled:
+			noPendingDecisionTask = false
+			buffer, history = flushBufferEvents(buffer, history)
+		case shared.EventTypeDecisionTaskCompleted:
+			noPendingDecisionTask = true
+			var completeBuffer []*shared.HistoryEvent
+			buffer, completeBuffer = getDecisionTaskCompletedBufferEvents(buffer)
+			buffer, history = flushBufferEvents(completeBuffer, history)
+		case shared.EventTypeDecisionTaskFailed:
+			noPendingDecisionTask = true
+			buffer, history = flushBufferEvents(buffer, history)
+		default:
+			if !noPendingDecisionTask {
+				buffer, history = flushBufferEvents(buffer, history)
+			}
+		}
+	}
+	if len(buffer) > 0 {
+		buffer, history = flushBufferEvents(buffer, history)
+	}
+	return history
+}
+
+func flushBufferEvents(buffer []*shared.HistoryEvent, history []*shared.History) ([]*shared.HistoryEvent, []*shared.History) {
+	newBatch := &shared.History{
+		Events: append([]*shared.HistoryEvent(nil), buffer...),
+	}
+	history = append(history, newBatch)
+	if len(buffer) > 0 {
+		buffer = buffer[:len(buffer)-1]
+	}
+	return buffer, history
+}
+
+func getDecisionTaskCompletedBufferEvents(buffer []*shared.HistoryEvent) (truncatedBuffer []*shared.HistoryEvent, completeBuffer []*shared.HistoryEvent) {
+	index := 0
+	for index < len(buffer) {
+		index++
+		if shouldBatchWithDecisionTaskCompleted(buffer[index]) {
+			completeBuffer = append(completeBuffer, buffer[index])
+		}
+		break
+	}
+	buffer = buffer[:index]
+	truncatedBuffer = buffer
+	return
+}
+
+func shouldBatchWithDecisionTaskCompleted(event *shared.HistoryEvent) bool {
+	switch *event.EventType {
+	case shared.EventTypeWorkflowExecutionCompleted,
+		shared.EventTypeWorkflowExecutionContinuedAsNew,
+		shared.EventTypeWorkflowExecutionFailed,
```

Comment:
missing EventTypeWorkflowExecutionCanceled

Reply:
The batch logic moved inside the generator. https://github.com/uber/cadence/pull/2280